### PR TITLE
Send messages using the (websocket) message pipe

### PR DESCRIPTION
**Pull request by gferon**
This adds the ability to submit messages to the `MessagePipe` (resolving #25).

Included is a new wrapper struct called `MessageToSend` which is deliberately not named `Message` as I thought it was too generic, any other name suggestion is appreciated.

From a high-level API view, it should look like:
```rust
let message_pipe = MessagePipe::from_socket(websocket, stream, credentials, message_sender);

// sender implements `Sink` and receiver implements `Stream` 
let (sender, receiver) = message_pipe.split();
```

TODO:
- [ ] Finish factoring the code to avoid duplication between `sender.rs` and `messagepipe.rs` _or_ remove sync message sending abilities altogether? Paging Dr. @rubdos
- [ ] Figure out how group messages should be sent (related to #75):
  1. the caller is responsible
  2. provide a helper method, like the one we have currently (`send_message_to_group`)
  3. something else
- [ ] Figure out if/why messages arrive only on the primary device, per @boxdot report when testing
- [ ] Maybe figure out the best way to have the `MessagePipe` work without having to poll on the receiving stream in order for sending to work... I suppose the obvious reflex would be to use `tokio::spawn`, which is not really available here. From a pure API point of view, we might want to `impl Future` on `MessagePipe` but then the user would be responsible to poll on 3 futures in order for the whole thing to work. I'm torn.

Should help @boxdot with https://github.com/boxdot/gurk-rs/issues/50

### CODE REVIEW

File: libsignal-service/src/messagepipe.rs [hunk 1]
```diff
@@ -119,9 +146,11 @@ impl<WS: WebSocketService> MessagePipe<WS> {
     }
 
     /// Worker task that
-    async fn run(
+    async fn run<'a>(
```
Comment (rubdos) [inline libsignal-service/src/messagepipe.rs:149]: Unused lifetime?

**Comment by Be-ing** [comment]
@nanuc uses a WebSocket to communicate between the Go backend of Axolotl and the Vue.js frontend. Might this be relevant to potentially using libsignal-service-rs in Axolotl?

**Comment by rubdos** [comment]
> @nanu-c uses a WebSocket to communicate between the Go backend of Axolotl and the Vue.js frontend. Might this be relevant to potentially using libsignal-service-rs in Axolotl?

No, this is using the websocket between the Signal servers and the client. Sounds like Axolotl uses a separate websocket between the client and the UI.

**Comment by rubdos** [comment]
> This adds the ability to submit messages to the `MessagePipe` (resolving #25).
> 
> Included is a new wrapper struct called `MessageToSend` which is deliberately not named `Message` as I thought it was too generic, any other name suggestion is appreciated.

`QueuedMessage`? That would imply that the `MessagePipe` implements some kind of smart-ish queueing system though.

>     * [ ]  Finish factoring the code to avoid duplication between `sender.rs` and `messagepipe.rs` _or_ remove sync message sending abilities altogether? Paging Dr. @rubdos

Nah, wouldn't drop it. Not everyone wants to use the full-fledged pipe I think. Maybe we can do some smart traits and wrappers to abstract over both?

>     * [ ]  Figure out how group messages should be sent (related to #75):
>       
>       1. the caller is responsible
>       2. provide a helper method, like the one we have currently (`send_message_to_group`)
>       3. something else

I would keep the sender and pipe in sync as much as possible, so possibly an abstraction over both solves this too.

>     * [ ]  Maybe figure out the best way to have the `MessagePipe` work without having to poll on the receiving stream in order for sending to work... I suppose the obvious reflex would be to use `tokio::spawn`, which is not really available here. From a pure API point of view, we might want to `impl Future` on `MessagePipe` but then the user would be responsible to poll on 3 futures in order for the whole thing to work. I'm torn.

`MessagePipe` already implements `Stream`, so you can poll whatever you need in there, no?

**Comment by rubdos** [comment]
> > MessagePipe already implements Stream, so you can poll whatever you need in there, no?
> 
> The problem I see here, is that somebody uses `.split()` and then submits a message in the `Sender` without polling on the `Stream` and nothing will happen. I don't think it's a massive problem though, just feels like it'd be easy to miss.

Well, those who don't use the stream should use the bare sender, no? You can also implement it in the `Sink` part, of course.

**Pull request closed without merging**